ca/certificate-authority.go

```diff
@@ -57,51 +59,104 @@ func NewCertificateAuthorityImpl(hostport string, authKey string, profile string
+xilAU/SLRXqt06pbv/hshGAJhl3F8Rs1/8gf2CorAw==
+-----END RSA PRIVATE KEY-----
+`
+
+// Given a CSR provided by the user and validated by the registration authority,
+// extract the names and public key, and pass them on to a CFSSL instance for
+// signing. IMPORTANT: This method must only be called by the registration
+// authority, and only after verifying that all names are authorized for the
+// requesting account key.
+// TODO: Add technical controls to only accept messages from the registration
+// authority.
+func (ca *CertificateAuthorityImpl) IssueCertificate(unsafeCsr x509.CertificateRequest) (cert core.Certificate, err error) {
+	// Only sign RSA certificates until we explicitly support ECDSA.
+	if unsafeCsr.PublicKeyAlgorithm != x509.RSA {
+		err = errors.New("Non-RSA key provided in CSR.")
+	}
+	// Don't accept the provided CSR as-is: Take only the public key and the names
+	// from it.
+	safeCsr := x509.CertificateRequest{
+		PublicKeyAlgorithm: unsafeCsr.PublicKeyAlgorithm,
+		PublicKey: unsafeCsr.PublicKey,
+	}
 	// Pull hostnames from CSR
-	hostNames := csr.DNSNames // DNSNames + CN from CSR
-	var commonName string
-	if len(csr.Subject.CommonName) > 0 {
-		commonName = csr.Subject.CommonName
+	hostNames := unsafeCsr.DNSNames // DNSNames + CN from CSR
+
+	// The CommonName from the Subject must always be present in the
+	// SubjectAltNames. If it was present, accept the provided CN.
+	cn := unsafeCsr.Subject.CommonName
+	if len(cn) > 0 {
+		hostNames = append(hostNames, cn)
+		safeCsr.Subject = pkix.Name{
+			CommonName: cn,
+		}
 	} else if len(hostNames) > 0 {
-		commonName = hostNames[0]
+		// If CN was not provided, pick one of the SubjectAltNames.
+		safeCsr.Subject = pkix.Name{
+			CommonName: hostNames[0],
+		}
 	} else {
+		// If there was no CN and no SubjectAltNames, fail.
 		err = errors.New("Cannot issue a certificate without a hostname.")
 		return
 	}
 
-	if len(hostNames) == 0 {
-		hostNames = []string{commonName}
-	}
-
-	identifier := core.AcmeIdentifier{Type: core.IdentifierDNS, Value: commonName}
-	if err = ca.PA.WillingToIssue(identifier); err != nil {
-		err = errors.New("Policy forbids issuing for name " + commonName)
-		return
-	}
+	// There may be duplicate entries in the input subjectAltNames + CN, so
+	// de-duplicate them with a map at the same time we check them for validity.
+	hostNamesMap := make(map[string]int)
 	for _, name := range hostNames {
-		identifier = core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name}
-		if err = ca.PA.WillingToIssue(identifier); err != nil {
+		identifier := core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name}
+		if willingErr := ca.PA.WillingToIssue(identifier); willingErr != nil {
 			err = errors.New("Policy forbids issuing for name " + name)
 			return
 		}
+		hostNamesMap[name] = 1
+	}
+
+	// Iterate through the map extracting the names again.
+	for k, _ := range hostNamesMap {
+		safeCsr.DNSNames = append(safeCsr.DNSNames, k)
```

Comment:
In order to prevent `append` from triggering resizes which cause memory allocations, the slice can be allocated with a capacity specified in advance:

```go
safeCsr.DNSNames = make([]string, 0, len(hostNamesMap))
```